### PR TITLE
[SPARK-38775][ML] cleanup validation functions

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?
1, remove unused `extractInstances` and `extractLabeledPoints` in `Predictor`;
2, remove unused `checkNonNegativeWeight` in `function`;
3, move `getNumClasses` from `Clasifier` to `DatasetUtils`;
4, move `getNumFeatures` from `MetadataUtils` to `DatasetUtils`;

### Why are the changes needed?
to unify to methods

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
existing testsuites

### CODE REVIEW

File: mllib/src/main/scala/org/apache/spark/ml/util/DatasetUtils.scala [hunk 1]
```diff
@@ -138,4 +140,61 @@ private[spark] object DatasetUtils {
...
+        numClasses
+    }
+  }
+
+  /**
+   * Obtain the number of features in a vector column.
+   * If no metadata is available, extract it from the dataset.
+   */
+  private[ml] def getNumFeatures(dataset: Dataset[_], vectorCol: String): Int = {
+    MetadataUtils.getNumFeatures(dataset.schema(vectorCol)) match {
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/ml/util/DatasetUtils.scala:193]: getOrElse works here too but doesn't matter

**Review by srowen — APPROVED** [review]
This is just refactoring - shouldn't change behavior? seems OK

**Review by dongjoon-hyun — COMMENTED** [review]
Hi, @zhengruifeng and @srowen (also cc @gengliangwang )

If you don't mind, could you hold on this refactoring a little bit more until @MaxGekk releases RC1?

1. Currently, Apache Spark 3.3 is under active testing and unstable.2. 
2. GitHub Action PR builder also doesn't provide you the ANSI test result
3. This kind of massive refactoring increases the complexity during the Apache Spark 3.3.0 release process.
   - For example, SPARK-38490 (https://github.com/apache/spark/pull/35797) introduced a ANSI conf related bug and [SPARK-38776](https://github.com/apache/spark/pull/36051) fixed it. However, SPARK-38669 (https://github.com/apache/spark/pull/35983) refactored `master` and `branch-3.3` different a lots and the backporting was difficult.

Hopefully, please give us more time until we finish Apache Spark 3.3, @zhengruifeng . I believe Cleanups and refactoring PRs are less urgent than the scheduled release.

File: mllib/src/main/scala/org/apache/spark/ml/util/DatasetUtils.scala (hunk 1 above)
Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/ml/util/DatasetUtils.scala:193]: ok, will swith back to getOrElse

**Comment by zhengruifeng** [comment]
@dongjoon-hyun  Ok, I will hold on this PR since its target version is 3.4

**Comment by dongjoon-hyun** [comment]
Thank you so much!

**Comment by srowen** [comment]
We can revisit this now. Rerun tests I think?

**Comment by dongjoon-hyun** [comment]
+1 for restarting. Thank you, @zhengruifeng and @srowen .

**Comment by zhengruifeng** [comment]
Sure, let me update this PR

**Review by huaxingao — APPROVED** [review]

**Review by dongjoon-hyun — APPROVED** [review]
+1, LGTM. Thank you, @zhengruifeng , @srowen , @huaxingao .
Merged to master.

**Comment by zhengruifeng** [comment]
Thanks all!

**Pull request closed without merging**